ggml/src/ggml-opencl2/ggml-opencl2.cpp

```diff
@@ -0,0 +1,3927 @@
+    /* .free                    = */ ggml_backend_opencl2_free,
+    /* .set_tensor_async        = */ NULL,  /* ggml_backend_opencl2_set_tensor_async */
+    /* .get_tensor_async        = */ NULL,  /* ggml_backend_opencl2_get_tensor_async */
+    /* .cpy_tensor_async        = */ NULL,  /* ggml_backend_opencl2_cpy_tensor_async */
+    /* .synchronize             = */ NULL,  /* ggml_backend_opencl2_synchronize */
+    /* .graph_plan_create       = */ NULL,
+    /* .graph_plan_free         = */ NULL,
+    /* .graph_plan_update       = */ NULL,
+    /* .graph_plan_compute      = */ NULL,
+    /* .graph_compute           = */ ggml_backend_opencl2_graph_compute,
+    /* .event_record            = */ NULL,
+    /* .event_wait              = */ NULL,
+};
+
+ggml_backend_t ggml_backend_opencl2_init(void) {
+    ggml_backend_dev_t dev = ggml_backend_reg_dev_get(ggml_backend_opencl2_reg(), 0);
+    ggml_backend_opencl2_context *backend_ctx = ggml_cl2_init(dev);
+
+    ggml_backend_t backend = new ggml_backend {
+        /* .guid      = */ ggml_backend_opencl2_guid(),
+        /* .interface = */ ggml_backend_opencl2_i,
+        /* .device    = */ dev,
+        /* .context   = */ backend_ctx
+    };
+
+    return backend;
+}
+
+bool ggml_backend_is_opencl2(ggml_backend_t backend) {
+    return backend && backend->iface.get_name == ggml_backend_opencl2_name;
+}
+
+//
+// buffer
+//
+struct ggml_backend_opencl2_buffer_context {
+    // A buffer context can hold multiple cl_mem objects. This is for flattening
+    // quantized weights and should be used with GGML_OPENCL_SMALL_ALLOC where
+    // each tensor is allocated a separate buffer. When flattening is enabled
+    // with small allocation, each tensor is backed by two cl_mem objects (for
+    // quants and scales) packed into a backend_opencl2_buffer.
+    ggml_backend_opencl2_buffer_context(cl_mem buf)
+        : name("OpenCL2") {
+        buffer.push_back(buf);
+    }
+
+    ~ggml_backend_opencl2_buffer_context() {
+        for (cl_mem buf : buffer) {
+            CL_CHECK(clReleaseMemObject(buf));
+        }
+        for (cl_mem im : img) {
+            CL_CHECK(clReleaseMemObject(im));
+        }
+
+        delete[] temp_tensor_extras;
+        delete[] temp_tensor_extras_q4_0;
+    }
+
+    ggml_tensor_extra_cl * ggml_opencl2_alloc_temp_tensor_extra() {
+        if (temp_tensor_extras == nullptr) {
+            temp_tensor_extras = new ggml_tensor_extra_cl[GGML_CL_MAX_NODES];
+        }
+
+        size_t alloc_index = temp_tensor_extra_index;
+        temp_tensor_extra_index = (temp_tensor_extra_index + 1) % GGML_CL_MAX_NODES;
+        ggml_tensor_extra_cl * extra = &temp_tensor_extras[alloc_index];
+        extra->reset();
+
+        return extra;
+    }
+
+    ggml_tensor_extra_cl_q4_0 * ggml_opencl2_alloc_temp_tensor_extra_q4_0() {
+        if (temp_tensor_extras_q4_0 == nullptr) {
+            temp_tensor_extras_q4_0 = new ggml_tensor_extra_cl_q4_0[GGML_CL_MAX_NODES];
+        }
```

Comment:
The number of available extras should not have a limit, otherwise it will fail when the number of tensors is higher. We have a better way to handle this with the `reset` function in the buffer interface. For example, this a possible implementation:
- Add vectors of available and used extras to the buffer context
- In `init_tensor`, take one from the available vector, or create a new one if empty. Add it to the vector of used extras.
- In `reset`, move all extras in the used vector to the available vector

Reply:
Fixed.